array.go

```diff
@@ -913,3 +913,43 @@ func countElement(array []interface{}, element interface{}) int {
 	}
 	return count
 }
+
+func (a *Array) Filter(filter func(index int, value *Value) bool) *Array {
+	a.chain.enter("Filter()")
+	defer a.chain.leave()
+
+	if a.chain.failed() {
+		return a
+	}
+
+	if filter == nil {
+		a.chain.fail(AssertionFailure{
+			Type: AssertUsage,
+			Errors: []error{
+				errors.New("unexpected nil function argument"),
+			},
+		})
+	}
+
+	var filteredArray []interface{}
+
+	for index, element := range a.value {
+		chainFailed := false
+		valueChain := a.chain.clone()
+		valueChain.setFatal(false)
+
+		valueChain.setFailCallback(func() {
+			chainFailed = true
+
+		})
+
+		if filter(index, newValue(valueChain, element)) && !chainFailed {
+			filteredArray = append(filteredArray, element)
+		} else {
+			valueChain.replace("Filter[%v]", element)
+		}
```

Comment:
This piece looks strange:

* why do you call replace only if predicate failed?
* why do you call it after predicate? the idea is to replace chain path before passing chain to newValue, so that assertion made on value inside predicate would have replaced path

Instead, we can just write:

```go
valueChain.replace("Filter[%v]", element)

if filter(index, newValue(valueChain, element)) && !chainFailed {
    ...
```